lib/interpreters/periodic_visits.rb

```diff
@@ -371,36 +373,63 @@ def associate_skills(new_vehicle, vehicle_day_index)
-          if service.unavailable_days.include?(day) || vrp.vehicles.none?{ |v| v.available_at(day) }
-            day += 1
+    def possible_days_from_possible_day(service, possible_days, vehicles, lapse, options)
+      day = @schedule_start
+      nb_services_seen = 0
+      set = []
+      while day <= @schedule_end && nb_services_seen < service.visits_number
+        # it is possible that user provided last_performed_visit_date/day_index or filled first_possible_days field
+        # we assume if there are X values in first_possible_days then it corresponds to X first visits of this service
+        # user needs to provide nil value if he wants to skip one service when providing first_possible_days
+        day =
+          if lapse
+            [possible_days[nb_services_seen], day].compact.max
           else
-            service.first_possible_days += [day]
-            nb_services_seen += 1
-            day += service.minimum_lapse || 1
+            # case with maximum_lapse, but maximum_lapse is nil
+            @schedule_end - (service.visits_number - nb_services_seen - 1) * (service.minimum_lapse || 1)
           end
+
+        while (service.unavailable_days.include?(day) || vehicles.none?{ |v| v.available_at(day) } || # day is not available
+              nb_services_seen > 0 && day <= set[nb_services_seen - 1] ||
+              options[:first_possible_days][nb_services_seen] && options[:first_possible_days][nb_services_seen] > day) &&
+              day.between?(@schedule_start, @schedule_end)
+          day += options[:is_min_lapse] ? 1 : -1
         end
 
-        # last possible day
-        day = @schedule_end
-        nb_services_seen = 0
-        while day >= @schedule_start && nb_services_seen < service.visits_number
-          if service.unavailable_days.include?(day) || vrp.vehicles.none?{ |v| v.available_at(day) }
-            day -= 1
-          else
-            service.last_possible_days += [day]
+        return set unless day.between?(@schedule_start, @schedule_end)
+
+        set << day
+        nb_services_seen += 1
+        day += lapse || 1
+      end
+
+      set
+    end
+    
+    def compute_possible_days(vrp)
+      # for each of this service's visits, computes first and last possible day to be assigned
+      vrp.services.each{ |service|
+        service.first_possible_days =
+          possible_days_from_possible_day(service, service[:first_possible_days], vrp.vehicles,
+                                          service.minimum_lapse || 1, { is_min_lapse: true, first_possible_days: [] })
+        if service.last_possible_days.any?
+          service.last_possible_days =
+            possible_days_from_possible_day(service, service[:last_possible_days], vrp.vehicles, service.maximum_lapse,
+                                            { is_min_lapse: false, first_possible_days: service[:first_possible_days] })
+        else
+          day = @schedule_end
+          nb_services_seen = 0
```

Comment:
```suggestion
          visit_index = 0
```

Moreover, we are are looking after every visit indices, a each loop seems more appropriate ?